### PR TITLE
Support fetching binary data using xhr.responseType = 'blob'

### PR DESCRIPTION
**Pull request by lrowe**
Perhaps it should request with `xhr.responseType = 'arraybuffer'` and convert to a string manually in `text()`.

### CODE REVIEW

**Comment by lrowe** [comment]
This is failing due to ariya/phantomjs#11013 which is only fixed for phantomjs 2.0.

File: fetch.js [hunk 1]
```diff
@@ -68,58 +68,85 @@
...
+      var match = charset_re.exec(this.headers.get('Content-Type') || '')
+      var charset = match ? match[1] : null
       var body = this._body
       return new Promise(function(resolve, reject) {
-        try {
-          resolve(JSON.parse(body))
-        } catch (ex) {
-          reject(ex)
+        var fileReader = new FileReader()
+        fileReader.readAsText(body, charset)
```
Comment (josh) [inline fetch.js]: I think only utf-8 decoding `text()` will be supported.

> Resolve p with the result of running utf-8 decode on bytes.

https://fetch.spec.whatwg.org/#body-mixin

**Comment by josh** [comment]
I'm a little worried about requiring XHR2 features like `responseType`. I don't want to jettison IE9 completely. I think the `.blob` mixin is the only one thats going to support binary data, text and json should still work fine given the utf-8 only constrain.

**Comment by lrowe** [comment]
Perhaps the implementation should switch based on `'responseType' in (new XMLHttpRequest()` leaving IE8/9 using responseText?
The workarounds for accessing binary data in ie8 look pretty nasty: http://miskun.com/javascript/binary-file-loading-in-javascript/. It might be possible to make something work combined with a typedarray polyfill like: https://bitbucket.org/lindenlab/llsd/src/default/js/typedarray.js but I'm not sure its worth the hassle.

**Comment by lrowe** [comment]
I've removed the charset handling per spec and use feature detection on 'responseType' to switch between implementations. All tests except for the already failing 308 redirect pass on IE9.

**Comment by josh** [comment]
> Perhaps the implementation should switch based on 'responseType' in (new XMLHttpRequest() leaving IE8/9 using responseText?

Yeah, that sounds pretty good to me. Don't worry about IE8 though.

File: fetch.js [hunk 2]
```diff
@@ -194,6 +226,9 @@
       }
 
       xhr.open(self.method, self.url)
+      if (has_responseType) {
```
Comment (josh) [inline fetch.js]: I'd probably just do the feature detect inline `'responseType' in xhr`

File: test/test.js [hunk 3]
```diff
@@ -63,6 +63,36 @@ suite('Response', function() {
 
 // https://fetch.spec.whatwg.org/#body-mixin
 suite('Body mixin', function() {
+  ;(Response.prototype.blob ? suite : suite.skip)('arrayBuffer', function() {
```
Comment (josh) [inline test/test.js]: Can we feature detect `Response.prototype.arrayBuffer` here?

File: test/server.js [hunk 4]
```diff
@@ -28,6 +28,10 @@ var routes = {
     });
     res.end('hi');
   },
+  '/nonascii': function(res) {
```
Comment (josh) [inline test/server.js]: Maybe just return some binary bytes from this test endpoint? Maybe even something thats invalid utf-8 just for good coverage.

File: fetch.js [hunk 5]
```diff
@@ -73,60 +73,92 @@
...
-        } catch (ex) {
-          reject(ex)
-        }
+      return this.text().then(function(text) {
+        return new Promise(function(resolve, reject) {
+          try {
+            resolve(JSON.parse(text))
+          } catch (ex) {
+            reject(ex)
+          }
```
Comment (josh) [inline fetch.js]: You don't need to make a new Promise here.

Just do `this.text().then(function(text) { return JSON.parse(text) })`

File: fetch.js [hunk 6]
```diff
@@ -73,60 +73,92 @@
...
-        return rejected ? rejected : Promise.resolve(decode(this._body))
+        return this.text().then(function(text) {
+          return new Promise(function(resolve, reject) {
+            try {
+              resolve(decode(text))
+            } catch (ex) {
+              reject(ex)
+            }
+          })
+        })
```
Comment (josh) [inline fetch.js]: ditto, `this.text().then(function(text) { return decode(text) })`

File: fetch.js [hunk 7]
```diff
@@ -73,60 +73,92 @@
     this._body = null
     this.bodyUsed = false
 
-    this.arrayBuffer = function() {
-      throw new Error('Not implemented yet')
-    }
+    if (has_responseType) {
```
Comment (josh) [inline fetch.js]: For IE9, checking for `FileReader` support should cover this definitions.

**Comment by josh** [comment]
> I'm not sure what to do about phantomjs.

Does phantom return a false positive for the `responseType` feature check? Trying to avoid the phantom UA checks, but I think some of the other checks should force it into the IE9 code path.

**Comment by josh** [comment]
@lrowe thanks for getting this going. Definitely a good direction here for handling binary data.

File: fetch.js [hunk 8]
```diff
@@ -73,60 +73,92 @@
...
+      this.arrayBuffer = function() {
```
Comment (josh) [inline fetch.js]: Maybe array buffer mixin is `ArrayBuffer && FileReader`

File: fetch.js [hunk 9]
```diff
@@ -73,60 +73,92 @@
...
+            } else {
+              resolve(e.target.result)
+            }
+          }
+        })
       }
-    })();
 
-    if (blobSupport) {
       this.blob = function() {
```
Comment (josh) [inline fetch.js]: We'll still want a blob feature check.

**Comment by lrowe** [comment]
> Does phantom return a false positive for the responseType feature check? Trying to avoid the phantom UA checks, but I think some of the other checks should force it into the IE9 code path.

Looks like its Blob type is broken but is otherwise based on a fairly recent WebKit.

File: fetch.js (hunk 2 above)
Comment (lrowe) [inline fetch.js]: Happy to change this, but note the module scope feature detection would still be required for switching method implementations in `Body`.

**Comment by josh** [comment]
> Looks like its Blob type is broken but is otherwise based on a fairly recent WebKit.

Yeah, thats why I had the blob constructor feature check. `var blobSupport = (function() {`

**Pull request merged**